### ISSUE TITLE
termination of a Process fails with collect and filter

### ISSUE THREAD

**Issue by alissapajer** [scalaz/scalaz-stream#290]
Hi folks,

What follows is a minimized case of a `Process` constructed with `collect` that doesn't halt when instructed to do so with a `wye.interrupt`.

```scala
import scalaz._
import scalaz.concurrent._
import scalaz.stream._
import scalaz.stream.async.mutable._

class Example {
  val poison: Signal[Boolean] = async.signal[Boolean]

  val original: Process[Task, Int] = Process.constant(42) //onComplete Process.eval_(Task.delay(println(s"ORIGINAL COMPLETE")))

  val secondary: Process[Task, Int] = original collect {
    case in if in < 0 => in
  } //onComplete Process.eval_(Task.delay(println(s"SECONDARY COMPLETE")))

  val output: Task[Seq[Int]] = {
    val signal: Process[Task, Boolean] = poison.discrete
    val result: Process[Task, Int] = (signal wye secondary)(wye.interrupt) //onComplete Process.eval_(Task.delay(println(s"INTERRUPT COMPLETE")))
    result.runLog[Task, Int]
  }

  object Effects extends Thread {
    override def run(): Unit = {
      poison.set(true).run
      println("POISON SET")
    }
  }
}

// val example = new Example(); example.Effects.start(); example.output.run.size
// for { i <- (1 to 100) } yield { val example = new Example(); example.Effects.start(); example.output.run.size }
```

When this example is run, the process sometimes halts (as expected), but more often than not it fails to halt. Here is a [thread dump](https://gist.github.com/alissapajer/c8237dd846cd522ea84a) taken while the process is spinning.

****Ways the example can be modified so that the process still spins****

(1) Include a value caught by the collect:
```scala
val original: Process[Task, Int] = Process(-12) ++ Process.constant(42)
```
(2) Use `filter` instead of `collect`: 
```scala
val secondary: Process[Task, Int] = original filter { _ < 0 }
```

****Ways the example can be modified so that the process never spins****

(1) Bypass the `collect` entirely with
```scala
val secondary = original
```
(2) Replace `<` with `>` in the `collect` `case` statement
```scala
case in if in > 0 => in
```
(3) Explicitly keep everything that is collected using
```scala
case in => in
```
(4) Use `filter` with `>` instead of collect: 
```scala
val secondary: Process[Task, Int] = original filter { _ > 0 }
```

**Comment by alissapajer** [comment]
@djspiewak @radekm @pchlupacek  What we've determined is that `Process#step` is infinite looping. Specifically, we are appending to `Halt(End)` to `Halt(End)` ad infinitum. In order for `step` to terminate, we need an `Emit` or an `Await` or a `Halt` with an empty stack, none of which we obtain.

In `collect`, when the predicate fails to be met, we append a `halt` to our stream (via `flatMap`). In the bug example, we supply an infinite stream of values, none of which satisfy `collect`'s predicate.

Basically, we're stepping through an infinite stream, appending halts along the way, without ever obtaining a value, and so we step and append forever.

**Comment by alissapajer** [comment]
A debugging note:  Since `step` is used in many places, I copied and pasted a new identical version of `step` called `myStep`, added printlns in that, and then determined that the call to `p.step` (changed to `p.myStep`) inside of `runStep` inside of `runAsync` is the call to `step` that never terminates.

**Comment by djspiewak** [comment]
Random update here…

After some brain storming with @puffnfresh, I'm pretty sure the correct solution here is to extend the `Step` algebra such that it fully represents the partiality monad over `Append`.  Practically, this means that `Step` would allow for a "nullary emit", in addition to the "emit" and "await" cases it currently allows.  Basically, `Step` could thus represent a pure continuation.  Then, the `step` function would thread through `Step` on `Append(Halt, …)`, ensuring that proper interleaving and fairness would work correctly.  As a side benefit, this would allow the `Append` stack members to directly return a `Process`, rather than wrapping in `Trampoline`.

A minor optimization on this to avoid dragging append evaluation performance through the mud would be to attempt to drill through up to *n* `Append` consecutive append nodes in the `go` inner function of `step`, and only returning the nullary `Step` when *n* has been exceeded.  It's worth noting that this strategy has been explored in the context of Scheme trampolines in some older papers, and it usually yields some marginal performance benefits over naive single-step trampolining, depending on your stack vs heap performance.

**Comment by puffnfresh** [comment]
:+1: 

Embedding the partiality monad would guarantee productivity in an infinite chain of Appends, which we can't have at the moment with `HaltOrStep`.

**Comment by pchiusano** [comment]
I'm sorry I'm so behind on responding to this. Thanks for everyone doing great work. I think it's safe to say that at the moment I have the least understanding of this issue out of anyone on this thread. :) This is my way of asking if someone wants to step up and try putting together a fix. :)

@djspiewak or @puffnfresh could you elaborate on:

> I'm pretty sure the correct solution here is to extend the Step algebra such that it fully represents the partiality monad over Append

As in, just show me what `Step` would look like with what you are thinking.

> A minor optimization on this to avoid dragging append evaluation performance through the mud would be to attempt to drill through up to n Append consecutive append nodes in the go inner function of step, and only returning the nullary Step when n has been exceeded.

This is sort of tangential, but I experimented at one point with a trampoline type in Scala that did exactly that, basically using `n` stack frames before bailing back to the central loop. At least in Scala, it ended up being a lot slower. My theory at the time is that constructing stack frames and tearing them down is pretty expensive on the JVM, and the GC has to traverse more stack to determine what's live. Whereas bump-a-counter generational is always pretty fast.

In any case, I'd say just implement the simplest thing that can possibly work, and we can optimize later if needed. :)

**Comment by djspiewak** [comment]
@pchiusano Very naively, `Step` could become the following:

```scala
case class Step[+F[_], +O](head: Option[EmitOrAwait[F, O]], next: Cont[F, O]) extends HaltOrStep[F, O] {
```

The `Option` gives `step` the ability to effectively say "uh, I don't have a value yet, but maybe I will later!" rather than just chewing on things indefinitely with an unfulfillable promise to always produce a value.

**Comment by pchiusano** [comment]
Gotcha. That makes sense.

**Comment by pchlupacek** [comment]
Not sure how much refactoring through the codebase this requires. 

We need to experiment a bit with it.

I also remember we run into issues with fact that step actually handles `Halt(End)` equally with `Emit(Nil)`. As such maybe we can just use that to signal `maybe will have something`. We once had `val Empty = Emit(Nil)` and maybe that is exactly  like `None` here. 

i mean perhaps removing just this line from step 
` case Emit(os) if os.isEmpty => go(Try(stack.head(End).run), stack.tail)`
can solve the current problem? 

Otherwise the `F` gives me a perfect sense in step, and I think that makes step more natural to me, and perhaps can be used in corner situations where you can use F ~> G to combine steps from different monads, perhaps (like new tcp.nio).

**Comment by radekm** [comment]
Making `head` of `Step` optional can guarantee productivity but will it guarantee progress? In which circumstances will be `head` `None`?

The possible problem is that between two calls of `step` a clutter may be added to the process (eg. by `suspend`s), so to do some real work `step` must at first remove this clutter.

**Comment by djspiewak** [comment]
> Making head of Step optional can guarantee productivity but will it guarantee progress? In which circumstances will be head None?

Progress cannot be guaranteed.  Give up hope.  Specifically:

```scala
lazy val derp: Process[Task, Unit] = Process.halt ++ derp
```

What do you *expect* `derp.step` to do?  I sure as heck don't know.  The types say there's a `Unit` value in there *somewhere*, but there's no way for `step` to ever get to it.  Progress as you seem to be defining it is simply not something that can be guaranteed with a turing complete system.

The only thing we can do is ensure that the `step` function allows safe interleaving of the computation.  Thus, while `derp` may run forever, it at least allows us to get other stuff done in the interim.

**Comment by radekm** [comment]
> Progress as you seem to be defining it is simply not something that can be guaranteed with a turing complete system.

Ok. What I meant was that if there is a value or an effect reachable by the current implementation in a finite number of steps then it should be finitely reachable in a new implementation.

Eg. always returning `Step` with `head == None` after `3` (or some fixed `n`) iterations of `go` in `step` will break this property since IMO you can construct a process where `suspend`s add more clutter than `step` is able to remove (so clutter accumulates at the beginning of the process and no value or effect is ever reached). On the other hand increasing `n` with each call to `step` would be IMO sufficient for most processes.

**Comment by djspiewak** [comment]
> Ok. What I meant was that if there is a value or an effect reachable by the current implementation in a finite number of steps then it should be finitely reachable in a new implementation.

This property would not be affected by a correct implementation of nullable `step`.  The only difference is that, under some circumstances, some of the steps which had been iterations of `go` would be outsourced to the external `step` driver.  The same iterations would exist though.

> On the other hand increasing n with each call to step would be IMO sufficient for most processes.

Side effects!  :-)

**Comment by pchiusano** [comment]
This is just an idea, but how about just changing `step` to take an `Option[Int]` for the maximum number of rounds. It sounds like `wye` is the only thing that would need to pass anything other than `None` (unbounded). It can pass increasing numbers in order to preserve productivity and fairness.

**Comment by djspiewak** [comment]
That's kind of what I was thinking. There needs to be some feedback mechanism though in order to let wye know whether or not the throttle was hit.  Also, it should probably start with a higher number, back off and then walk back up again to tune, depending on fairness.

**Comment by pchlupacek** [comment]
Are you guys sure this affects only wye?

**Comment by radekm** [comment]
> It sounds like `wye` is the only thing that would need to pass anything other than None (unbounded).

I think that all `step`s called directly or indirectly from `wye` will need the limit. So it will be necessary to propagate it everywhere (eg. to `pipe` which uses `step`).

**Comment by pchiusano** [comment]
Yikes, that's going to be annoying. :\ Can anyone think of another approach
or is this basically the only option?

**Comment by pchlupacek** [comment]
I think I have the solution that does not require introducing the int. 

First, we need to modify step as follows  (note commented lines to be removed)

```scala

final def step: HaltOrStep[F, O] = {
    def go(cur: Process[F,O], stack: Vector[Cause => Trampoline[Process[F,O]]]) : HaltOrStep[F,O] = {
      if (stack.nonEmpty) cur match {
        case Halt(cause) => go(Try(stack.head(cause).run), stack.tail)
      //  case Emit(os) if os.isEmpty => go(Try(stack.head(End).run), stack.tail)
        case emt@(Emit(os)) => Step(emt,Cont(stack))
        case awt@Await(_,_) => Step(awt,Cont(stack))
        case Append(h,st) => go(h, st fast_++ stack)
      } else cur match {
        case hlt@Halt(cause) => hlt
      //  case emt@Emit(os) if (os.isEmpty) => halt0
        case emt@Emit(os) => Step(emt,Cont(Vector.empty))
        case awt@Await(_,_) => Step(awt,Cont(Vector.empty))
        case Append(h,st) => go(h,st)
      }
    }
    go(this,Vector.empty)

  }
```

then we need to modify runAsync as follows :

(line 1337)
```scala

   //@tailrec
          def runStep(p: Process[Task, O]): Cause \/ (EarlyCause => Process[Task,O]) = {
            val step = p.step
            step match {
             // case Step(Emit(Seq()), cont)         => runStep(cont.continue)
              case Step(Emit(h), cont)             => S(cb(right((h, cont)))); left(End)
              case Step(awt@Await(req, rcv), cont) =>
                req.runAsync(r => a ! AwaitDone(r, awt, cont))
                right((c:EarlyCause) => rcv(left(c)).run +: cont)
              case Halt(cause)                 => S(cb(left(cause))); left(cause)
            }
          }

```

then we introduce 

```scala 
  val empty = Emit(Nil)
```

and need to change collect to 

```scala 

def collect[I, I2](pf: PartialFunction[I, I2]): Process1[I, I2] =
    id[I].flatMap(pf andThen (emit) orElse { case _ => empty /* was halt */ })

```
then maybe I would suggest to deprecate `halt` or perhaps just define 

```scala 
 @deprecated
  val halt = Emit(Nil)
 // type where we really must signal Halt(End) for effectivity (like `onHalt`)
  val halt0 = Halt(End)
``` 

I don't like to introducing the int. It will require another level of thought when using runAsync or step throughout the code, specifically when user really is stepping for some low-level combinators. 

But, If, we really decide for int in step, I would suggest just to change signature of the step to be : 

```scala 

final def step(max:Int = 1000): HaltOrStep[F, O] = ???

```

if possible rather than poluting the step type.

**Comment by radekm** [comment]
@pchlupacek This indeed fixes the test

```scala
  property("interrupt-constant.signal-halt.collect-none") = secure {
    val p1 = Process.constant(42).collect { case i if i < 0 => i }
    val i1 = Process(false)
    val v = i1.wye(p1)(wye.interrupt).runLog.timed(3000).run.toList
    v.size >= 0
  }
```

given by @alissapajer but if you replace `p1` by `p2` defined as

```scala
val p2 = p1 |> process1.id
```

then it is broken again.

**Comment by pchlupacek** [comment]
@radekm  you think even in case when 

```val halt == Emit(Nil)```

???

**Comment by pchlupacek** [comment]
Personally I think we have to make the `step` to exit at every iteration when Emit(Nil) is encountered. Then, in various combinators (like wye) eventually optimize (go) so we won't emit to central actor unless needed (ech nth, on nonempty, halt etc). That will cause the `int` to be closed in implementation of all actors. and won't introduce too much overhead? What do you think? This in fact will as well maybe allow us to leave step as it is and perhaps as well leave the halt as it is.  

Perhaps, becuase runAsync itself id backed by actor, maybe we can just solve this in runAsync?

**Comment by radekm** [comment]
When I set `val halt = Emit(Nil)` then `wye` doesn't even terminate. On the other hand if I let `halt` untouched (i.e. `val halt = halt0`) then `wye` terminates but the process `p2` still runs in background.

**Comment by alissapajer** [comment]
After a conversation with @djspiewak, I added
```scala
case Append(Halt(End), st) => Step(Emit(Nil), Cont(st))
```
to the case of `step` when the stack is empty, the idea being that `Emit(Nil)` can represent the "I don't have a value to emit, but I might later" type of the Partiality.

This addition causes the following `Process` to fail to halt:
```scala
Process.range(0, 10).toSource.runLog.run
```
What's happening is, the `step` function is relying on recursing with `Halt(End)` and a nonempty stack in order to get a further `Await` or `Emit`. So thus `step` is relying on actually running the function at the head of `stack` with the cause `End` as an input. When we bail out of step early with the `Emit(Nil)`, we never get a chance to actually run the function on the stack.

Intuitively, we need to jump out of the recursion to give the `Emit(Nil)` (or whatever) a chance to propagate, and then, if necessary, jump back into `step` in the same state we were in before. 

This presents the question: once out `step`, how do we further signal to a `Wye`, say, that we didnt' get a value but still might. This got me thinking about the `ReceiveY` algebra, and how, to break out of an `await` that's never going to complete, we may need to add `ReceiveNothing` value.

**Comment by pchlupacek** [comment]
@alissapajer, @radekm @pchiusano , @djspiewak,  @puffnfresh I have did some tests and I think I have found solution to this issue. 

First, I think @radekm was right pointing out that in certain cases `step` shall never terminate. I believe that this is correct behaviour of step that it will never terminate with these processes : 

```scala

Process.constant(42).collect { case i if i < 0 => i }

Process.constant(42).filter { _ < 0 }
 ```

and similar. 

Another observation is that `step` is actually used only in `pure` computations, such as pipe, tee and wye (meaning wye internally).  The only place where step was used for async computation was in runAsync, like @radekm discovered. 

The runAsync was actually the root cause of this problem as it reused step and unfortunately in such a way that for non-terminating processes (like above) blocked fully the thread that was supposed to be asynchronous, doing the job on `other` thread, and caused combinators like wye to never leave runAsync by blocking runAsync invocation with never-terminating processes.

I had made changes to runAsync, and all tests specified by @alissapajer now passing. Can you folks take a look on the branch (https://github.com/scalaz/scalaz-stream/tree/bug/step-runAsync) and see if this fixes the issue with all relevant cases you may think of? I may be improving the runAsync implementation before merge, but would like to get feedback before if the fix is correct.

Thanks a lot to all for finding out this issue and helping out.

**Comment by alissapajer** [comment]
@pchlupacek The case that @radekm suggested earlier using `pipe` does not halt, when it should.

So, specifically, this does not halt on @pchlupacek 's latest bug-fix branch `bug/step-runAsync`:
```scala
val p1 = Process.constant(42).collect { case i if i < 0 => i }
val p2 = p1 |> process1.id
val i1 = Process(false)
val v = i1.wye(p2)(wye.interrupt).runLog.timed(3000).run.toList
```
It's my understanding that `pipe`ing through `process1.id` should never change any behavior. But if you instead run the following (replacing `p2` with `p1`) 
```scala
val v = i1.wye(p1)(wye.interrupt).runLog.timed(3000).run.toList
```
on `bug/step-runAsync`, it will indeed terminate, as expected (and as it should).

`println`s in `Process#pipe` show that we're infinitely emitting `42`, and never halting. In the `match this.step` we match on `Step(Emit(Vector(42)),Cont(...))`, which recurses back into `pipe`. The `Process1` input to `pipe` is never halting, so that `suspendStep` is never giving us a `Halt`.

/cc @djspiewak

**Comment by pchlupacek** [comment]
@alissapajer I'll take look on that case.  

Actually the way how runAsync is implemented now is that we don't care if process is infinite and won't emit. It will always get interrupted. In this case it shall be interrupted because `i1` terminated. Whats likely to be happening is either that wye don't sent the kill signal or there is a problem with Process.kill on 
`p2`. Would be interested if `p2.kill` terminates. 

On the process1.id, I am actually not sure if it is well defined. 
perhaps we have to implement it as (can you pls try)?

```scala

def id[A] : Process1[A,A] = receive1( a => emit(a) ++ id)

```

**Comment by pchlupacek** [comment]
Hm I think I see whats happening. The problem is that we still don't have chance to interrupt piped processes that run internally the Step. I may need to think about this for a while.

**Comment by alissapajer** [comment]
Yes, that makes sense. The process is infinite as expected, but we don't have a chance to interrupt it.

I changed the `id` impl to
```scala
def id[A] : Process1[A,A] = receive1( a => emit(a) ++ id)
```
and the problem persists.

Interestingly, when one changes `v` to
```scala
val v = i1.wye(p2.kill)(wye.interrupt).runLog.timed(3000).run.toList
```
the process does terminate.

**Comment by pchlupacek** [comment]
@alissapajer can you perhaps try with this step implementation? : 

```scala 

  final def step: HaltOrStep[F, O] = {
    @tailrec
    def go(cur: Process[F,O], stack: Vector[Cause => Trampoline[Process[F,O]]], cnt: Int) : HaltOrStep[F,O] = {
      if (stack.nonEmpty) cur match {
        case Halt(End) if cnt <=0  => Step(empty,Cont(stack))
        case Halt(cause) => go(Try(stack.head(cause).run), stack.tail, cnt -1)
        case Emit(os) if os.isEmpty => Step(empty,Cont(stack))
        case emt@(Emit(os)) => Step(emt,Cont(stack))
        case awt@Await(_,_) => Step(awt,Cont(stack))
        case Append(h,st) => go(h, st fast_++ stack, cnt -1)
      } else cur match {
        case hlt@Halt(cause) => hlt
        case emt@Emit(os) if (os.isEmpty) => halt0
        case emt@Emit(os) => Step(emt,Cont(Vector.empty))
        case awt@Await(_,_) => Step(awt,Cont(Vector.empty))
        case Append(h,st) => go(h,st, cnt -1)
      }
    }
    go(this,Vector.empty, 10)

  }



```

**Comment by pchlupacek** [comment]
ah sorry and define emtpy as : 
  /** Empty process that never emits anything */
  val empty: Emit[Nothing] = Emit(Nil)

**Comment by alissapajer** [comment]
@pchlupacek with that change to `step`, the `pipe`d process terminates! So giving ourselves a way out fo the `step` recursion (however we want to do that in the end) allows a chance to interrupt the process.

Additionally, I ran a full `sbt test` on the `bug/step-runAsync` branch, and everything passed.

Though, I'm not convinced that a process isn't still running in the background. I'm running `console` in `sbt`, and when I `:quit` out of the console, it just hangs, and never quits. Here's a [thread dump](https://gist.github.com/alissapajer/b2806e96671afe5e0db8) from after the example terminates. It's troublesome that the console isn't quitting, but none of the threads are running `scalaz` code, so I'm not sure what's up.

**Comment by pchlupacek** [comment]
@alissapajer To me the stack trace is unrelated to our code. I have same observation with or w/o fix, But lets write a test that will verify there is no dangling process in background. 

I am happy that we now see the solution. I will think about proper step implementation again to eventually remove that internal count and define back runAsync in term of step. 

I'll try to finish that over weekend.

**Comment by pchiusano** [comment]
Awesome!

**Comment by alissapajer** [comment]
FYI, branch [bug/step-runAsync](https://github.com/alissapajer/scalaz-stream/tree/bug/step-runAsync) pushed to my fork contains the broken pipe test as well as the "fix" which adds a counter to `step`, in case anyone needs to build locally against it to work around the bug for now.

**Comment by pchlupacek** [comment]
@alissapajer cool, excellent. That gives me more time to think how to fix this properly.

**Comment by djspiewak** [comment]
Fixed now with the landing of bug/step(!!!!!!)
